In [115]:
import numpy as np
from scipy.optimize import minimize
from scipy.io import loadmat
from math import sqrt
import matplotlib.pyplot as plt
import random
%matplotlib inline

#  Loading training data

In [116]:
mat = loadmat('./mnist_all.mat')

In [ ]:
mat.keys()

In [119]:
# get all the training data: 
print 'training data:'
train0 = mat.get('train0')
print train0.shape
train1 = mat.get('train1')
print train1.shape
train2 = mat.get('train2')
print train2.shape
train3 = mat.get('train3')
print train3.shape
train4 = mat.get('train4')
print train4.shape
train5 = mat.get('train5')
print train5.shape
train6 = mat.get('train6')
print train6.shape
train7 = mat.get('train7')
print train7.shape
train8 = mat.get('train8')
print train8.shape
train9 = mat.get('train9')
print train9.shape
print train_mat.shape

# get all the testing data: 
print 'testing data:'
test0 = mat.get('test0')
print test0.shape
test1 = mat.get('test1')
print test1.shape
test2 = mat.get('test2')
print test2.shape
test3 = mat.get('test3')
print test3.shape
test4 = mat.get('test4')
print test4.shape
test5 = mat.get('test5')
print test5.shape
test6 = mat.get('test6')
print test6.shape
test7 = mat.get('test7')
print test7.shape
test8 = mat.get('test8')
print test8.shape
test9 = mat.get('test9')
print test9.shape

training data:
(5923, 784)
(6742, 784)
(5958, 784)
(6131, 784)
(5842, 784)
(5421, 784)
(5918, 784)
(6265, 784)
(5851, 784)
(5949, 784)
(60000, 784)
testing data:
(980, 784)
(1135, 784)
(1032, 784)
(1010, 784)
(982, 784)
(892, 784)
(958, 784)
(1028, 784)
(974, 784)
(1009, 784)


# (1) Stack all training matrices into one 60000  784 matrix. Do the same for test matrices.

In [120]:
train_mat = np.concatenate((train0, train1, train2, train3, train4, 
                            train5, train6, train7, train8, train9), 
                            axis=0)
print 'training data dimension:'
print train_mat.shape
print type(train_mat)

test_mat = np.concatenate((test0, test1, test2, test3, test4, 
                            test5, test6, test7, test8, test9), 
                            axis=0)
print 'training data dimension:'
print test_mat.shape
print type(test_mat)

training data dimension:
(60000, 784)
<type 'numpy.ndarray'>
training data dimension:
(10000, 784)
<type 'numpy.ndarray'>


# (2) Create a 60000 length vector with true labels (digits) for each training example. Same for test data.

In [124]:
# training data : 
shape_train = [train0.shape[0], train1.shape[0],train2.shape[0],train3.shape[0],
         train4.shape[0], train5.shape[0],train6.shape[0],train7.shape[0],
         train8.shape[0], train9.shape[0]]
print shape_train
shape_train = np.array(shape_train)

training_data_list = []
labels = 0;

for shape in shape_train:
    array = np.ones((shape,), dtype=np.int)   
    array = labels*array
    training_data_list.append(array)
    labels = labels + 1;

train_label_all = []
for classes in training_data_list:
    train_label_all = np.concatenate((train_label_all, classes), axis=0)
print train_label_all
print train_label_all.shape

# test data : 
shape_test = [test0.shape[0], test1.shape[0],test2.shape[0],test3.shape[0],
         test4.shape[0], test5.shape[0],test6.shape[0],test7.shape[0],
         test8.shape[0], test9.shape[0]]
print shape_test
shape_test = np.array(shape_test)

test_data_list = []
labels = 0;

for shape in shape_test:
    array = np.ones((shape,), dtype=np.int)   
    array = labels*array
    test_data_list.append(array)
    labels = labels + 1;

test_label_all = []
for classes in test_data_list:
    test_label_all = np.concatenate((test_label_all, classes), axis=0)
print test_label_all
print test_label_all.shape

[5923, 6742, 5958, 6131, 5842, 5421, 5918, 6265, 5851, 5949]
[ 0.  0.  0. ...,  9.  9.  9.]
(60000,)
[980, 1135, 1032, 1010, 982, 892, 958, 1028, 974, 1009]
[ 0.  0.  0. ...,  9.  9.  9.]
(10000,)


# (3) Normalize the training matrix and test matrix so that the values are between 0 and 1.

In [126]:
# training data : 
normalized_train_mat = (train_mat / 255)
print 'training matrix:'
print normalized_train_mat 
#test data : 
normalized_test_mat = (test_mat / 255)
print 'test matrix'
print normalized_test_mat 

training matrix:
[[0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 ..., 
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]]
test matrix
[[0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 ..., 
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]
 [0 0 0 ..., 0 0 0]]


# Randomly split the 60000 X 784 normalized matrix into two matrices: training matrix (50000 X 784) and validation matrix (10000 X  784). Make sure you split the true labels vector into two parts as well.

In [132]:
# marge normalized training matrix and training label
train_mat_with_label = np.zeros((60000,785))
train_mat_with_label[:,:-1] = normalized_train_mat
train_mat_with_label[:,-1] = train_label_all
print train_mat_with_label
print 'matrix dimension: '
print train_mat_with_label.shape

[[ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 ..., 
 [ 0.  0.  0. ...,  0.  0.  9.]
 [ 0.  0.  0. ...,  0.  0.  9.]
 [ 0.  0.  0. ...,  0.  0.  9.]]
matrix dimension: 
(60000, 785)


In [134]:
# shuffle rows randomly 
np.random.shuffle(train_mat_with_label)
print train_mat_with_label

[[ 0.  0.  0. ...,  0.  0.  4.]
 [ 0.  0.  0. ...,  0.  0.  3.]
 [ 0.  0.  0. ...,  0.  0.  5.]
 ..., 
 [ 0.  0.  0. ...,  0.  0.  4.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  7.]]


In [136]:
# perform split :
train_data = train_mat_with_label[:50000,:-1]
train_label = train_mat_with_label[:50000,-1]
validation_data = train_mat_with_label[50000:60000,:-1]
validation_label = train_mat_with_label[50000:60000,-1]
print train_data.shape
print train_label.shape
print validation_data.shape
print validation_label.shape

(50000, 784)
(50000,)
(10000, 784)
(10000,)


In [137]:
# test data : 
test_data = normalized_test_mat
test_label = test_label_all
print test_data.shape
print test_label.shape

(10000, 784)
(10000,)
